semantic-reranker/src/main/java/org/opensearch/search/relevance/actionfilter/SearchActionFilter.java

```diff
@@ -0,0 +1,370 @@
+  }
+
+  private boolean shouldDoSemanticRerank(final SearchRequest searchRequest) {
+    if (searchRequest.source() == null) {
+      return false;
+    }
+
+    // Skip if there is scroll, pagination, or sorting.
+    if (searchRequest.scroll() != null || searchRequest.source().from() > 0 ||
+        (searchRequest.source().sorts() != null && !searchRequest.source().sorts().isEmpty())) {
+      return false;
+    }
+
+    final String[] indices = searchRequest.indices();
+    // Skip if no or more than 1 indices is specified.
+    if (indices == null || indices.length != 1) {
+      return false;
+    }
+
+    // Check request-level setting
+    if (!searchRequest.source().ext().isEmpty()) {
+      // Filter ext builders by name
+      List<SearchExtBuilder> extBuilders = searchRequest.source().ext().stream()
+          .filter(searchExtBuilder -> KendraSearchExtBuilder.NAME.equals(searchExtBuilder.getWriteableName()))
+          .collect(Collectors.toList());
+      if (!extBuilders.isEmpty()) {
+        KendraSearchExtBuilder kendraSearchExtBuilder = (KendraSearchExtBuilder) extBuilders.get(0);
+        return kendraSearchExtBuilder.isRankerEnabled();
+      }
+    }
+
+    // Check index level setting
+    Settings settings = openSearchClient.getIndexSettings(indices[0], new String[] { Constants.ENABLED_SETTING_NAME });
+    // Skip if plugin enabled flag is not true.
+    if (settings == null || !TRUE.equals(settings.get(Constants.ENABLED_SETTING_NAME))) {
+      return false;
+    }
+    return true;
+  }
+
+  private <Response extends ActionResponse> ActionListener<Response> createSearchResponseListener(
+      final ActionListener<Response> listener,
+      final long startTime,
+      final QueryParser.QueryParserResult queryParserResult,
+      final boolean pluginEnabledFetchSource) {
+    return new ActionListener<Response>() {
+
+      @Override
+      public void onResponse(final Response response) {
+        final SearchResponse searchResponse = (SearchResponse) response;
+        final long totalHits = searchResponse.getHits().getTotalHits().value;
+        if (totalHits == 0) {
+          LOGGER.info("TotalHits = 0. Returning search response without semantic reranking: {}", searchResponse);
+          listener.onResponse(response);
+          return;
+        }
+
+        LOGGER.info("Starting semantic reranking for search response: {}", searchResponse);
+        
+        try {
+          final BytesStreamOutput out = new BytesStreamOutput();
+          searchResponse.writeTo(out);
+
+          final StreamInput in = new NamedWriteableAwareStreamInput(out.bytes().streamInput(), namedWriteableRegistry);
+
+          LOGGER.info("Extracting search hits");
+          final SearchHits hits = new SearchHits(in);
+
+          final SearchHits newHits = doSemanticRerank(queryParserResult, hits, pluginEnabledFetchSource);
+
+          LOGGER.info("Extracting remaining response fields");
```

Comment:
I feel like this section can (and should) be extracted to support any external reranker. Anyone who is trying to manipulate the `SearchHits` of a `SearchResponse` would need to jump through these same hoops.

More generally, how hard would it be to split out the generic reranking logic from the Kendra-specific implementation? 

I think the generic logic is probably:
1. Call an abstract `shouldRerank` method based on the search request, and if true, create a generic response listener with a `Reranker`.
2. The generic response listener extracts the `SearchHits` from the response, passes them to the reranker, gets back new `SearchHits` and uses them (and the old `SearchResponse`) to create a new `SearchResponse`.

IMO, that would also help for breaking up this class (which is a little big right now).

Reply:
Thanks for the suggestion, I've refactored in the latest commit.